### PR TITLE
[HACKATHON] Improve performance on many concurrent uploads

### PR DESCRIPTION
**Pull request by cosa65**
# Background
#### Link to staging deployment URL 
https://ui-martinfosco-ui541.scp-staging.biomage.net/

#### Anything else the reviewers should know about the changes here

# Changes
### Code changes

# Definition of DONE
Your changes will be ready for merging after each of the steps below have been completed:

### Code updates
Have best practices and ongoing refactors being observed in this PR

- [ ] Migrated any selector / reducer used to the new format
- [ ] Validated that current unit tests work as expected and are enough

### Testing
- [ ] Unit tests written
- [ ] Tested locally with Inframock (with latest production data downloaded with biomage experiment pull)
- [ ] Deployed to staging
- [ ] Passed integration tests 

To set up easy local testing with inframock, follow the instructions here: https://github.com/biomage-ltd/inframock
To deploy to the staging environment, follow the instructions here: https://github.com/biomage-ltd/biomage-utils

### Documentation updates
Is all relevant documentation updated to reflect the proposed changes in this PR?

- [ ] Relevant Github READMEs updated
- [ ] Relevant wiki pages created/updated

### Approvers
- [ ] Approved by a member of the core engineering team
- [ ] (UX changes) Approved by vickymorrison (this is her username, tag her if you need approval)

### Just before merging:
- [ ] After the PR is approved, the `unstage` script in here: https://github.com/biomage-ltd/biomage-utils is executed. This script cleans up your deployment to staging

### Optional
- [ ] Photo of a cute animal attached to this PR

### CODE REVIEW

File: src/components/data-management/SamplesTableCells.jsx [hunk 1]
```diff
@@ -35,9 +37,16 @@ const UploadCell = (props) => {
...
-  const { progress = null, status = null } = file?.upload ?? {};
+  const { progressEmitter = null, status = null } = file?.upload ?? {};
   const [uploadDetailsModalVisible, setUploadDetailsModalVisible] = useState(false);
+  const [uploadProgress, setUploadProgress] = useState(null);
+
   const uploadDetailsModalDataRef = useRef(null);
+  useEffect(() => {
+    // Disabled because eslint is dumb and doesn't recognize function calls if they have "?" before
+    // eslint-disable-next-line no-unused-expressions
+    progressEmitter?.on('progress', setUploadProgress);
```
Comment (aerlaut) [inline src/components/data-management/SamplesTableCells.jsx]: by doing `progressEmitter?` are you basically saying `if(progressEmitter) progressEmitter.on(...)`?

File: src/redux/actions/samples/deleteSamples.js [hunk 2]
```diff
@@ -39,9 +39,10 @@ const sendDeleteSamplesRequest = async (projectUuid, experimentId, sampleUuids)
 
 const cancelUploads = async (files) => {
   const promises = Object.values(files).map(({ upload }) => {
-    if (upload?.amplifyPromise) {
-      // return Storage.cancel(upload.amplifyPromise);
-    }
+    // Disabled because eslint is dumb and doesn't recognize function calls if they have "?" before
+    // eslint-disable-next-line no-unused-expressions
+    upload?.cancelToken.cancel();
```
Comment (aerlaut) [inline src/redux/actions/samples/deleteSamples.js]: by doing `upload?` are you basically saying `if(upload) upload.on(...)?`

File: src/redux/reducers/samples/initialState.js [hunk 3]
```diff
@@ -24,7 +24,8 @@ const sampleFileTemplate = {
   lastModified: '',
   upload: {
     status: null,
-    amplifyPromise: null,
+    progressEmitter: null,
```
Comment (aerlaut) [inline src/redux/reducers/samples/initialState.js:27]: this is not saved to the database right?

File: src/components/data-management/SamplesTableCells.jsx (hunk 1 above)
Comment (cosa65) [reply src/components/data-management/SamplesTableCells.jsx]: Yes, it's a common pattern that eslint doesn't recognize for some reason

File: src/redux/actions/samples/deleteSamples.js (hunk 2 above)
Comment (cosa65) [reply src/redux/actions/samples/deleteSamples.js]: Same as here https://github.com/biomage-ltd/ui/pull/541#discussion_r739254093

File: src/redux/reducers/samples/initialState.js (hunk 3 above)
Comment (cosa65) [reply src/redux/reducers/samples/initialState.js:27]: amplifyPromise was being saved to the database, and progress emitter will probably be saved too

Comment (cosa65) [reply src/redux/reducers/samples/initialState.js:27]: maybe I can do some cleanup on this here, you're right

File: src/utils/upload/processUpload.js [hunk 4]
```diff
@@ -28,7 +28,10 @@ const putInS3 = async (projectUuid, loadedFileData, dispatch, sampleUuid, fileNa
 
   const signedUrl = await signedUrlResponse.json();
 
-  return axios.request({
+  const progressEmitter = new EventEmitter();
+  const cancelToken = axios.CancelToken.source();
```
Comment (aerlaut) [inline src/utils/upload/processUpload.js]: Don't you have to assign this `cancelToken` to a request?

**Review by aerlaut — DISMISSED** [review]
LGTM 👍

**Pull request closed without merging**